### PR TITLE
SystemProfiler: ensure samples are collected from event

### PR DESCRIPTION
**Pull request by roi-granulate**
## Checklist:
* [ ] Please post on the PR an example log displaying how it'd look when this new check fail (you can inject a fault to the code to make sure it does break, for example remove the "ok" return)
* [ ] Sanity test on aarch64

### CODE REVIEW

File: gprofiler/utils/perf_process.py
Comment (roi-granulate) [inline gprofiler/utils/perf_process.py:48]: I could choose a better name I guess

File: gprofiler/profiler_state.py
Comment (Jongy) [inline gprofiler/profiler_state.py:12]: nitpick - swap them, `if TYPE CHECKING` imports are nicer after real imports.

File: gprofiler/profilers/perf.py
Comment (Jongy) [inline gprofiler/profilers/perf.py:250]: This check is actually "is perf enabled". Nicer to check as `if not disabled` and I think that we already have the `not in ("none", "disabled")` check somewhere - please share.

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:29]: `storage_dir` is the term used across the project, otherwise use `tmp_dir`, don't add new terms

File: gprofiler/utils/perf_process.py
Comment (Jongy) [inline gprofiler/utils/perf_process.py:30]: You removed it, but is it not relevant anymore?

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:55]: Add some explanation on what's happening here, in short. "We've observed that on some machines, the default event used by perf is never triggered and hence no samples are collected ...."

**Review by Jongy — COMMENTED** [review]
Not done reviewing

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py]: do note we pass `storage_dir` for this function
but in the scope of this function it is indeed `tmp_dir`
544e75ff8883d6de5fc5ad339a79497e345b5863

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py:58]: I find docstrings sufficient, anyways, please review this suggested change
3e7d22623c0f0d93a97dd5cc4fdb82017fc5ef9c

File: gprofiler/profilers/perf.py
Comment (roi-granulate) [reply gprofiler/profilers/perf.py]: gotcha, please note the `cast` - I think it's safe but anyways I can just accept `int | bool | str | None` (`user_args.get` ret type) in `is_profiler_disabled`
f370b148ff1e80657f395d4044fa25fe530cd008

File: gprofiler/utils/perf_process.py
Comment (roi-granulate) [reply gprofiler/utils/perf_process.py]: 18de021cf661320184a695bd12f43b489735f8dd

File: gprofiler/profilers/perf.py
Comment (Jongy) [inline gprofiler/profilers/perf.py:251]: Huh? If perf mode is none/disabled then we don't want to run any perf at all, what is this check for?

File: gprofiler/profilers/perf.py
Comment (Jongy) [reply gprofiler/profilers/perf.py]: Cast is ok

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:36]:
```suggestion
    actually collects samples, and make changes only if it doesn't.
```

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:23]:
```suggestion
    def perf_extra_args(self) -> List[str]:
```

File: gprofiler/profilers/perf.py
Comment (roi-granulate) [reply gprofiler/profilers/perf.py:251]: yea it should have been ```suggestion
        if not is_profiler_disabled(perf_mode):
```

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:56]: Need a matching `perf_process.stop()`, no? Works nicely in `finally` block

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py:23]: it can return empty list tho when `self == SUPPORTED_PERF_EVENTS.PERF_DEFAULT`
so `List`

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:19]: Order is important here, because you want to test `default` first, right? Please comment it then (so no-one changes the order)

File: gprofiler/utils/perf_process.py
Comment (Jongy) [inline gprofiler/utils/perf_process.py:49]: Instead of adding a new arg - we can ensure the `extra_args` are last, and then the extra args become `[ -e, event, --, sleep, 0.5 ]`

File: gprofiler/profilers/perf.py
Comment (Jongy) [inline gprofiler/profilers/perf.py:257]: You should exit after this, no? raise a critical error that'll cause gprofiler to exit.

File: gprofiler/utils/perf_process.py
Comment (Jongy) [reply gprofiler/utils/perf_process.py:62]: That's fine

**Review by Jongy — CHANGES_REQUESTED** [review]
Done reviewing

File: gprofiler/profilers/perf.py
Comment (roi-granulate) [reply gprofiler/profilers/perf.py:277]: 15ecd0d4256c9a2bacd1b803c23280a1f9235f14

File: gprofiler/profilers/perf.py
Comment (roi-granulate) [reply gprofiler/profilers/perf.py]: 63b1be26ae67e5177b94768c73485637fb78f9f0

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py]: 252e6b55e19ea771544e269eeec817fb14cbb1c6

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py]: b789ea97cedfff3cc34c3a60784c214a875e083a

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py:57]: 8b81c1ae1423af3fc9ad99feb38d4a30bc13d80f

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py:26]: please review
68b59a2cb3458d430c8fe6a1953c0979ed732557

File: gprofiler/utils/perf_process.py
Comment (roi-granulate) [reply gprofiler/utils/perf_process.py]: I prefer the current approach
but idc e86ad7b4d28bcfe1caff7eb2f0493469bce6d9e7

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py]: b9a04a5c37b5154220bb975ac4cff759acf41ecf

File: gprofiler/profilers/perf.py
Comment (roi-granulate) [reply gprofiler/profilers/perf.py:257]: how should we do that? can we just `exit(1)` when some profilers have already started?

Comment (roi-granulate) [reply gprofiler/profilers/perf.py:257]: talked f2f, it's ok to `sys.exit(1)` here
although we could just let the exception propagate couldn't we?

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:48]: Is this needed? (if mypy required it but you don't initialize - we'll just get UnboundLocalError).

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:36]: Please add some spacing in this function to make it clearer. For example - 
1. Blank between the docstring and code
2. blank between the try block and code in loop
3. blank between the try block and the final return line

File: gprofiler/profilers/perf.py
Comment (Jongy) [reply gprofiler/profilers/perf.py:258]: We can let the exception propagate. Please post on the PR an example log displaying how it'd look when this fail (you can inject a fault to the code to make sure it does break, for example remove the "ok" return)

File: gprofiler/profilers/perf.py
Comment (Jongy) [inline gprofiler/profilers/perf.py:252]: Is this check even needed? If perf is disabled - we never get into the c'tor of this profiler, right?

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:65]: Perhaps you should call `_parse_perf_script` here. The perf script can be logically empty (== 0 samples) but the text won't be empty - include blanks, comment lines etc that `_parse_perf_script` is stripping.

**Review by Jongy — CHANGES_REQUESTED** [review]
Done reviewing

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py]: definitely no, it's there due to old implementations I did locally 1b4e23ff1d00d959f382f7708070864f3324c007

File: gprofiler/profilers/perf.py
Comment (roi-granulate) [reply gprofiler/profilers/perf.py:258]: great, let's keep this unresolved so we'll remember to do it when I run this?

File: gprofiler/profilers/perf.py
Comment (roi-granulate) [reply gprofiler/profilers/perf.py]: 9f050e09ad2246e65b9ab3150ce310a3d3fb1e49

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py:66]: moved this function to utils in a different pr
can you please review?
https://github.com/Granulate/gprofiler/pull/889

File: gprofiler/profilers/perf.py
Comment (Jongy) [reply gprofiler/profilers/perf.py]: I resolved but added a checkbox task in the PR desc

File: gprofiler/utils/perf.py
Comment (Jongy) [reply gprofiler/utils/perf.py:66]: Approved, waiting for you to merge and update the code.

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:36]: Function name & first line of doscstring can be improved. The name "perf_default_event_works" does not describe an action - it's a statement. And the first line "Validate that `perf record`'s default event actually collects samples." is just incorrect. The function returns the list of arguments to use the appropriate event in perf record - please improve the name & docstring.

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:53]: Perhaps add a small doc on the `sleep 0.5` - the reason is to ensure we'll get at least a few samples.

**Review by Jongy — CHANGES_REQUESTED** [review]
Done reviewing

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py]: df01baee3145c5017910115cd78c08e41b11c732

Comment (roi-granulate) [reply gprofiler/utils/perf.py]: wdyt of `get_perf_event_args`?

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py]: 86732158d5a0f1cd2c8ec9671d7d9d04e1974a19

File: gprofiler/utils/perf.py
Comment (Jongy) [reply gprofiler/utils/perf.py]: Good

**Comment by roi-granulate** [comment]
@Jongy attaching the log of a failed run (I `not` this condition on a "healthy" machine)
```
$ sudo ./build/x86_64/gprofiler --output-dir=/tmp/
[2024-02-28 17:23:17,662] INFO: gprofiler: Running gProfiler (version=1.45.0, commandline=--output-dir=/tmp/, arguments={'pid_file': '/var/run/gprofiler.pid', 'config': None, 'frequency': 11, 'duration': 60, 'insert_dso_name': False, 'output_dir': '/tmp/', 'flamegraph': True, 'profiling_mode': 'cpu', 'alloc_interval': None, 'rotating_output': False, 'pids_to_profile': None, 'dotnet_mode': 'disabled', 'python_mode': 'auto', 'python_add_versions': True, 'python_pyperf_user_stacks_pages': None, 'python_pyperf_verbose': False, 'java_mode': 'ap', 'java_version_check': True, 'java_async_profiler_mode': 'auto', 'java_async_profiler_safemode': 0, 'java_async_profiler_features': ['probesp', 'vtable'], 'java_async_profiler_args': None, 'java_safemode': 'profiled-oom,profiled-signaled,hserr', 'java_jattach_timeout': 30, 'java_async_profiler_mcache': 30, 'java_collect_spark_app_name_as_appid': False, 'java_async_profiler_report_meminfo': True, 'java_collect_jvm_flags': 'default', 'java_full_hserr': False, 'java_include_method_modifiers': False, 'java_line_numbers': 'none', 'perf_mode': 'fp', 'perf_dwarf_stack_size': 8192, 'perf_memory_restart': True, 'php_mode': 'disabled', 'php_process_filter': 'php-fpm', 'ruby_mode': 'rbspy', 'spark_sample_period': 120, 'collect_spark_metrics': False, 'nodejs_mode': 'disabled', 'log_usage': False, 'upload_results': False, 'disable_core_files': True, 'subcommand': None, 'server_host': 'https://profiler.granulate.io', 'api_server': 'https://api.granulate.io', 'server_upload_timeout': 120, 'server_token': None, 'service_name': None, 'curlify_requests': False, 'verify': True, 'verbose': False, 'log_file': '/var/log/gprofiler/gprofiler.log', 'log_rotate_max_size': 5242880, 'log_rotate_backup_count': 1, 'log_to_server': True, 'container_names': True, 'continuous': False, 'profile_api_version': None, 'pid_ns_check': True, 'collect_metrics': True, 'collect_metadata': True, 'collect_appids': True, 'app_id_args_filters': [], 'application_metadata': True, 'controller_pid': None, 'external_metadata': None, 'databricks_job_name_as_service_name': False, 'profile_spawned_processes': False, 'diagnostics': False, 'perf_inject': False, 'perf_node_attach': False, 'extract_resources': False})
[2024-02-28 17:23:17,831] INFO: gprofiler: gProfiler Python version: 3.10.13 (main, Feb 21 2024, 15:25:32) [GCC 4.8.5 20150623 (Red Hat 4.8.5-44)]
[2024-02-28 17:23:17,832] INFO: gprofiler: gProfiler deployment mode: standalone_executable
[2024-02-28 17:23:17,832] INFO: gprofiler: Kernel uname release: 6.5.0
[2024-02-28 17:23:17,832] INFO: gprofiler: Kernel uname version: #14~22.04.1-Ubuntu SMP Thu Feb 15 15:27:06 UTC 2024
[2024-02-28 17:23:17,833] INFO: gprofiler: Total CPUs: 8
[2024-02-28 17:23:17,833] INFO: gprofiler: Total RAM: 31.33 GB
[2024-02-28 17:23:17,833] INFO: gprofiler: Linux distribution: Ubuntu | 22.04 | Jammy Jellyfish
[2024-02-28 17:23:17,833] INFO: gprofiler: libc version: glibc-2.35-0ubuntu3.6
[2024-02-28 17:23:17,834] INFO: gprofiler: Hostname: hostname
[2024-02-28 17:23:17,895] INFO: gprofiler.containers_client: Discovered container runtimes: ['docker']
[2024-02-28 17:23:17,896] INFO: gprofiler.profilers.profiler_base: Initialized PythonEbpfProfiler (frequency: 11hz, duration: 60s), profiling mode: cpu
[2024-02-28 17:23:24,134] INFO: gprofiler.profilers.profiler_base: Initialized JavaProfiler (frequency: 11hz, duration: 60s), profiling mode: cpu
[2024-02-28 17:23:28,619] INFO: gprofiler.profilers.profiler_base: Initialized SystemProfiler (frequency: 11hz, duration: 60s), profiling mode: cpu
[2024-02-28 17:23:28,620] INFO: gprofiler.utils.perf_process: Starting perf (fp mode)
[2024-02-28 17:23:28,726] INFO: gprofiler.utils.perf_process: Started perf (fp mode)
[2024-02-28 17:23:30,749] INFO: gprofiler.utils.perf_process: Stopped perf (fp mode) (exit_code=0, stderr=b'', stdout=b'')
[2024-02-28 17:23:30,749] INFO: gprofiler.utils.perf_process: Starting perf (fp mode)
[2024-02-28 17:23:30,856] INFO: gprofiler.utils.perf_process: Started perf (fp mode)
[2024-02-28 17:23:32,197] INFO: gprofiler.utils.perf_process: Stopped perf (fp mode) (exit_code=0, stderr=b'', stdout=b'')
[2024-02-28 17:23:32,198] INFO: gprofiler.utils.perf_process: Starting perf (fp mode)
[2024-02-28 17:23:32,304] INFO: gprofiler.utils.perf_process: Started perf (fp mode)
[2024-02-28 17:23:33,659] INFO: gprofiler.utils.perf_process: Stopped perf (fp mode) (exit_code=0, stderr=b'', stdout=b'')
[2024-02-28 17:23:33,659] CRITICAL: gprofiler.profilers.perf: Failed to determine perf event to use
[2024-02-28 17:23:33,660] CRITICAL: gprofiler.profilers.factory: Couldn't create the Perf profiler, not continuing. Run with --no-perf to disable this profiler
Traceback (most recent call last):
  File "gprofiler/profilers/factory.py", line 54, in get_profilers
    profiler_instance = profiler_config.profiler_class(**profiler_kwargs)
  File "gprofiler/profilers/perf.py", line 185, in __init__
    get_perf_event_args(Path(self._profiler_state.storage_dir), self._profiler_state.stop_event)
  File "gprofiler/utils/perf.py", line 104, in get_perf_event_args
    raise PerfNoSupportedEvent
gprofiler.exceptions.PerfNoSupportedEvent
```

File: gprofiler/profilers/perf.py
Comment (Jongy) [inline gprofiler/profilers/perf.py:184]: Let's log the chosen event (default / cpu-clock / cpu-cycles) here.

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:41]: nitpick - name it properly and remove the pylint. This is not C/C++. `class SupportedPerfEvent` is okay

File: gprofiler/utils/perf.py
Comment (Jongy) [inline gprofiler/utils/perf.py:79]: Note on using `PerfProcess` here - it's slightly wasteful and slows down gProfiler boot, because `PerfProcess` works with continuous perf (`--switch-output`). We don't require that for the `sleep 0.5` short run we're trying to do, it only makes things longer by 1.5s per my measuring.
Please take a look if you can perhaps add code in `PerfProcess` that'll allow to control a perf started without this option (control it with a c'tor parameter?), then we don't need the `start` and `stop` and it can be just a single call to run it which returns with the `perf script` result.
If it's a mess to add support - let's consider skipping it (it is not a must improvement IMO).

File: gprofiler/profilers/perf.py
Comment (roi-granulate) [reply gprofiler/profilers/perf.py:184]: we could change `get_perf_event_args` so that it will return `SUPPORTED_PERF_EVENTS` instead of `List`, and then log the event itself instead of the picked list.
I think that somewhat nicer, lmk what you think

Comment (roi-granulate) [reply gprofiler/profilers/perf.py]: f3cecb035fbcc92b524744f9539a81a3fd64022e

File: gprofiler/utils/perf.py
Comment (roi-granulate) [reply gprofiler/utils/perf.py:80]: what makes you believe that this makes a difference?
will `--switch-output=signal` still have such overhead impact? if it won't, I think this approach is fine 2d72a63286a97125b3b6365cacefee4fd4607b77

Comment (roi-granulate) [reply gprofiler/utils/perf.py:80]: what makes you believe that this makes a difference?
will --switch-output=signal still have such overhead impact? if it won't, I think this approach is fine 2d72a63286a97125b3b6365cacefee4fd4607b77

File: gprofiler/profilers/perf.py
Comment (Jongy) [inline gprofiler/profilers/perf.py:188]: discovered_perf_event.name (so it shows just the string?)

**Review by Jongy — APPROVED** [review]
No more comments. I just want to try it locally a bit, after which I'll merge it tmrw.

**Comment by Jongy** [comment]
Looks fine - I'll merge after CI with merged master passes, and we'll release 1.46.0 tmrw.

**Pull request merged**
